File: src/Scripting/Lua/patterns/memento.lua
local editor = { text = "draft" }
local function save(originator) return { text = originator.text } end
local function restore(originator, snapshot) originator.text = snapshot.text end
local snapshot = save(editor)
editor.text = "broken"
restore(editor, snapshot)
assert(editor.text == "draft")
return true

File: src/Scripting/PHP/patterns/memento.php
<?php declare(strict_types=1);
$check = static fn(bool $ok) => $ok ?: throw new RuntimeException('memento');
$originator = new class {
    public string $state = 'draft';
    public function save(): array { return ['state' => $this->state]; }
    public function restore(array $memento): void { $this->state = $memento['state']; }
};
$m = $originator->save(); $originator->state = 'published'; $originator->restore($m);
$check($originator->state === 'draft');
